cmd/argo/commands/copy_artifacts.go

```diff
@@ -0,0 +1,184 @@
+			//var artifactOutput v1alpha1.Artifacts
+			//Getting a list of Artifacts based on user input
+			if len(copyArtifactArgs.nodeId) > 0 && len(copyArtifactArgs.templateName) > 0 {
+				log.Fatal("Only one of --template-name and --node-id can be filled out")
+			}
+			var nodes v1alpha1.Nodes
+			if len(copyArtifactArgs.nodeId) > 0 {
+				nodes = workflow.Status.Nodes.Filter(func(n v1alpha1.NodeStatus) bool { return n.ID == copyArtifactArgs.nodeId })
+			} else if len(copyArtifactArgs.templateName) > 0 {
+				nodes = workflow.Status.Nodes.Filter(func(n v1alpha1.NodeStatus) bool { return n.TemplateName == copyArtifactArgs.templateName })
+			} else {
+				nodes = workflow.Status.Nodes
+			}
+			for _, node := range nodes {
+				if node.Outputs != nil {
+					artifactOutput[node.ID] = node.Outputs.Artifacts
+				}
+			}
+
+			//Filtering in case of specifying artifact name
+			keysToDelete := make([]string, 0)
+			if len(copyArtifactArgs.artifactName) > 0 {
+				for key, artifactList := range artifactOutput {
+					artifactPtr := artifactList.GetArtifactByName(copyArtifactArgs.artifactName)
+					if artifactPtr != nil {
+						artifactOutput[key] = v1alpha1.Artifacts{*artifactPtr}
+					} else {
+						keysToDelete = append(keysToDelete, key)
+					}
+				}
+				for _, key := range keysToDelete {
+					delete(artifactOutput, key)
+				}
+			}
+
+			outputDir := args[0]
+			if len(outputDir) == 0 {
+				log.Fatal("No output directory was specified")
+			}
+			basePath := filepath.Join(outputDir, namespace, copyArtifactArgs.workflowName)
+			err = os.MkdirAll(basePath, os.ModePerm)
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			c := makeClient(argoServerArgs)
+			for nodeId, artifactList := range artifactOutput {
+				nodeIdFolder := filepath.Join(basePath, nodeId, "outputs")
+				err = os.MkdirAll(nodeIdFolder, os.ModePerm)
+				if err != nil {
+					log.Fatal(err)
+				}
+				for _, artifact := range artifactList {
+					getAndStoreArtifactData(namespace, copyArtifactArgs.workflowName, nodeId, artifact.Name, nodeIdFolder, c, argoServerArgs)
+				}
+			}
+
+		},
+	}
+	command.Flags().StringVarP(&copyArtifactArgs.namespace, "namespace", "n", "", "namespace of workflow")
+	command.Flags().StringVar(&copyArtifactArgs.workflowName, "workflow-name", "", "name of workflow")
+	command.Flags().StringVar(&copyArtifactArgs.nodeId, "node-id", "", "id of node in workflow")
+	command.Flags().StringVar(&copyArtifactArgs.templateName, "template-name", "", "name of template in workflow")
+	command.Flags().StringVar(&copyArtifactArgs.artifactName, "artifact-name", "", "name of output artifact in workflow")
+	command.Flags().StringVar(&copyArtifactArgs.artifactType, "artifact-type", "output", "type of artifact to copy (only supports output as of now)")
+
+	command.PersistentFlags().StringVarP(&argoServerArgs.URL, "argo-server", "s", os.Getenv("ARGO_SERVER"), "API server `host:port`. e.g. localhost:2746. Defaults to the ARGO_SERVER environment variable.")
```

Comment:
is this already accessible in another variable? I think this is dupe

Reply:
They are duplicates from `conn.go`, the main roadblock here was that the `argoServerOpts` method that the values are added to is private, and cannot be accessed. I could make a method in `conn.go` to make the client/setup the base request if that would be smoother?